src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/IntervalTreeAlgorithms`2.cs

```diff
@@ -0,0 +1,158 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+
+namespace Microsoft.CodeAnalysis.Shared.Collections;
+
+/// <summary>
+/// Helpers for working with <see cref="IIntervalTree{T}"/> instances.  Can be retrieved by calling <c>.Extensions</c>
+/// on an interval tree instance.  This is exposed as a struct instead of extension methods as the type inference
+/// involved here is too complex for C# to handle (specifically using a <c>TIntervalTree</c> type), which would make
+/// ergonomics extremely painful as the callsites would have to pass three type arguments along explicitly.
+/// </summary>
+internal readonly struct IntervalTreeAlgorithms<T, TIntervalTree>(TIntervalTree tree) where TIntervalTree : IIntervalTree<T>
```

Comment:
all the algorithms that will work fine on any interval tree impl (binary, or flat array).  extracted out from the Binary version.